### DISCUSSION TITLE
Provide API docs and specify public API

### DISCUSSION THREAD

**Discussion by chriskrycho** [11ty/eleventy#4110] [Enhancement Queue]
**Is your feature request related to a problem? Please describe.**

Right now, there are two major limitations to Eleventy's API documentation:

1. Information on the APIs is scattered throughout guide material, making it difficult to find *all* the capabilities of the tool.
2. There is no single source of truth for what constitutes the public API of the project, i.e. what is safe to rely on and what is not.

This makes it difficult to know what parts of the existing implementation are safe to use and which aren't. It also has made it difficult to build accurate TypeScript type definitions for the project.

(Not everyone wants to use TypeScript, of course, and that's totally fine! For those who *do* want to use TS, though this is really important—and even for those who don't, it can *really* improve the dev experience for writing plain-old JavaScript. Anything which uses the TS language server, especially VS Code but equally something like Vim, can get benefits from this!)

**Describe the solution you'd like**

It would be *great* if Eleventy included dedicated API docs which specified the public API of the project, separate from the guides material. The guides are genuinely fabulous, but they constitute a different kind of docs than exhaustive API docs do. (As a point of reference, consider [this great writeup of different kinds of documentation](http://stevelosh.com/blog/2013/09/teach-dont-tell/)!)

**Describe alternatives you've considered**

The obvious alternative is to do nothing: leave the API documentation only available in the guides. This is definitely less work!

**Additional context**

An example of how this can improve dev experience:

<img width="883" alt="Screen Shot 2019-12-07 at 08 05 42" src="https://user-images.githubusercontent.com/2403023/70376661-88733980-18c8-11ea-9412-76e707394ded.png">

**Comment by zachleat** [comment]
This is even more important now with Eleventy Serverless.

**Comment by zachleat** [comment]
Oh one more comment here, I do want to make sure that this is published on the web site and not included in the bundle, for bundle size reasons https://github.com/11ty/eleventy/issues/1885

**Comment by panoply** [comment]
See - https://github.com/11ty/eleventy/issues/577#issuecomment-1126862063

Also Related:

- https://github.com/11ty/eleventy/issues/814
- https://github.com/11ty/eleventy/issues/2033

**Comment by zachleat** [comment]
Just for completeness I’d link up a very minimal page here https://www.11ty.dev/docs/programmatic/
